service/history/queue/transfer_queue_processor_base.go

```diff
@@ -255,67 +270,83 @@ processorPumpLoop:
 	}
 }
 
-func (t *transferQueueProcessorBase) processBatch() {
-	ctx, cancel := context.WithTimeout(context.Background(), loadQueueTaskThrottleRetryDelay)
-	if err := t.rateLimiter.Wait(ctx); err != nil {
-		cancel()
-		t.notifyNewTask()
-		return
-	}
-	cancel()
-
-	t.lastPollTime = t.shard.GetTimeSource().Now()
-
+func (t *transferQueueProcessorBase) processQueueCollections(levels map[int]struct{}) {
 	t.queueCollectionsLock.RLock()
 	processingQueueCollections := t.processingQueueCollections
 	t.queueCollectionsLock.RUnlock()
 
-	// TODO: create a feedback loop to slow down loading for non-default queues (queues with level > 0)
 	for _, queueCollection := range processingQueueCollections {
 		t.queueCollectionsLock.RLock()
+		level := queueCollection.Level()
+		if _, ok := levels[level]; !ok {
+			t.queueCollectionsLock.RUnlock()
+			continue
+		}
+
 		activeQueue := queueCollection.ActiveQueue()
 		if activeQueue == nil {
+			// process for this queue collection has finished
+			// it's possible that new queue will be added to this collection later though,
+			// pollTime will be updated after split/merge
 			t.queueCollectionsLock.RUnlock()
 			continue
 		}
 
 		readLevel := activeQueue.State().ReadLevel()
-		maxReadLevel := activeQueue.State().MaxLevel()
+		maxReadLevel := minTaskKey(activeQueue.State().MaxLevel(), t.updateMaxReadLevel())
+		domainFilter := activeQueue.State().DomainFilter()
 		t.queueCollectionsLock.RUnlock()
 
-		shardMaxReadLevel := t.updateMaxReadLevel()
-		if shardMaxReadLevel.Less(maxReadLevel) {
-			maxReadLevel = shardMaxReadLevel
-			if !readLevel.Less(maxReadLevel) {
-				continue
+		if !readLevel.Less(maxReadLevel) {
+			if level != defaultProcessingQueueLevel {
+				// we only need to notify non default queue in this case
+				// poll time for default queue will be updated through notifyNewTask
+				t.upsertPollTime(level, t.shard.GetTimeSource().Now().Add(nonDefaultQueueBackoffDuration))
 			}
+			continue
 		}
 
+		ctx, cancel := context.WithTimeout(context.Background(), loadQueueTaskThrottleRetryDelay)
+		if err := t.rateLimiter.Wait(ctx); err != nil {
+			cancel()
+			backoff := time.Duration(0)
+			if level != defaultProcessingQueueLevel {
+				backoff = nonDefaultQueueBackoffDuration
+			}
+			t.upsertPollTime(level, t.shard.GetTimeSource().Now().Add(backoff))
+			continue
+		}
+		cancel()
+
 		transferTaskInfos, more, err := t.readTasks(readLevel, maxReadLevel)
 		if err != nil {
 			t.logger.Error("Processor unable to retrieve tasks", tag.Error(err))
-			t.notifyNewTask() // re-enqueue the event
-			return
+			t.upsertPollTime(level, time.Time{}) // re-enqueue the event
+			continue
 		}
 
+		pollTime := t.shard.GetTimeSource().Now()
+		t.upsertPollTime(level, pollTime.Add(backoff.JitDuration(
```

Comment:
do you need to check if it is default level here?

Reply:
I think it makes sense to set a max poll interval for queues at all level just as a safety mechanism. Let me know your thoughts on this.

Reply:
I see. LGTM